crates/analyze/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+#![deny(unused_crate_dependencies)]
+use anyhow::{
+    anyhow,
+    bail,
+    Result,
+};
+pub use contract_metadata::Language;
+use parity_wasm::elements::Module;
+
+/// Detects the programming language of a smart contract from its WebAssembly (Wasm)
+/// binary code.
+///
+/// This function accepts a Wasm code as input and employs a set of heuristics to identify
+/// the contract's source language. It currently supports detection for Ink!, Solidity,
+/// and AssemblyScript languages.
+///
+/// If multiple language patterns are found in the code, the function returns an error.
+pub fn determine_language(code: &[u8]) -> Result<Language> {
+    let module: Module = parity_wasm::deserialize_buffer(code)?;
+    let import_section = module.import_section();
+    let start_section = module.start_section();
+    let mut custom_sections = module.custom_sections().map(|e| e.name()).peekable();
+
+    let import_section_first = import_section
+        .ok_or(anyhow!("Missing required import section"))?
+        .entries()
+        .first()
+        .map(|e| e.field())
+        .ok_or(anyhow!("Missing required import section"))?;
+
+    if import_section_first != "memory"
+        && start_section.is_none()
+        && custom_sections.peek().is_none()
```

Comment:
I'm not 100% convinced by this heuristic, I know it is copied from that python script. 

It is because it is identifying the language based on the combination of a *lack* of things, which has the potential for false positives.

I would prefer to match on the presence of something, where we can be reasonably sure that it is an `ink` contract. We could add a custom section like the other languages (although that excludes contracts built with older versions). Alternatively we could identify the signature of a commonly used function somehow? e.g. a quick glance at a few contracts I can spot something that similar to:

incrementer.wat
```
(func (;8;) (type 5) (param i32)
    (local i32 i32 i32 i32)
    global.get 0
    i32.const 16
    i32.sub
    local.tee 1
    global.set 0
    local.get 1
    i64.const 16384
    i64.store offset=4 align=4
    local.get 1
    i32.const 65536
    i32.store
    local.get 1
    i32.const 0
    call 9
    block  ;; label = @1
```
flipper.wat
```
(func (;11;) (type 6) (param i32)
    (local i32 i32 i32 i32)
    global.get 0
    i32.const 32
    i32.sub
    local.tee 1
    global.set 0
    local.get 1
    i64.const 16384
    i64.store offset=20 align=4
    local.get 1
    i32.const 65536
    i32.store offset=16
    local.get 1
    i32.const 0
    i32.store offset=12
    local.get 1
    i32.const 16
    i32.add
    local.get 1
    i32.const 12
    i32.add
    i32.const 4
    call 6
    block  ;; label = @1
````
Although not sure which function that corresponds to, but maybe there is another ink core library function we can find which is guaranteed to be included in the generated Wasm.

Anyway I would prefer to be more strict and miss some contracts that are written in ink rather than falsely categorising some contracts as having been written in `ink`.

Reply:
I'm not a fan of this heuristic too. I only included it because the task was defined such way.
I can take a look to see if it could be done in a better way.

Reply:
I think that basing on function presence can be risky, because inlining optimization can affect our logic.